lib/nmatrix/math.rb

```diff
@@ -1179,6 +1179,36 @@ def scale(alpha, incx=1, n=nil)
   alias :permute_columns  :laswp
   alias :permute_columns! :laswp!
 
+  #
+  #call-seq:
+  #    positive_semidefinite? -> boolean
+  #
+  # A matrix is positive semidefinite if all of its eigenvalues are non-negative.
+  #
+  # * *Returns* :
+  #   - A booloean value telling if the NMatrix is positive semidefinite or not.
+  # * *Raises* :
+  #   - +ShapeError+ -> Must be used on square matrices
+  #
+  def positive_semidefinite?
+    raised(ShapeError, "positive semidefinite calculated only for square matrices")unless
+      self.dim == 2 && self.shape[0] == self.shape[1]
+    # Cast to a dtype for which geev is implemented
+    new_dtype = self.integer_dtype? ? :float64 : self.dtype
+    copy = self.cast(:dense, new_dtype)
+    eigenvalues = NMatrix::LAPACK.geev(copy, :left)
+    ans = true
+    ind = 0
+    while ind != self.dim
+      if eigenvalues[0][ind].real < 0
+        ans = false
```

Comment:
So this is somewhat inefficient. You can tell the first time `ans` gets set to `false` that this matrix is going to fail, so you should probably just `return false` instead. You don't need `ans` at all; if it makes it to the end of the while loop without returning, it should `return true` or just `true` (if it's the last line in scope).